core/query.go

```diff
@@ -1,17 +1,291 @@
 	}
-	return query.Pipeline
+	q.clientCallContext[dgst] = &ClientCallContext{
+		Deps:   deps,
+		ModID:  modID,
+		FnCall: call,
+	}
+	return nil
+}
+
+func (q *Query) CurrentModule(ctx context.Context) (dagql.ID[*Module], error) {
+	var id dagql.ID[*Module]
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return id, err
+	}
+	if clientMetadata.ModuleCallerDigest == "" {
+		return id, fmt.Errorf("no current module for main client caller")
+	}
+
+	q.clientCallMu.RLock()
+	defer q.clientCallMu.RUnlock()
+	callCtx, ok := q.clientCallContext[clientMetadata.ModuleCallerDigest]
+	if !ok {
+		return id, fmt.Errorf("client call %s not found", clientMetadata.ModuleCallerDigest)
+	}
+	return dagql.NewID[*Module](callCtx.ModID), nil
+}
+
+func (q *Query) CurrentFunctionCall(ctx context.Context) (*FunctionCall, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if clientMetadata.ModuleCallerDigest == "" {
+		return nil, fmt.Errorf("no current function call for main client caller")
+	}
+
+	q.clientCallMu.RLock()
+	defer q.clientCallMu.RUnlock()
+	callCtx, ok := q.clientCallContext[clientMetadata.ModuleCallerDigest]
+	if !ok {
+		return nil, fmt.Errorf("client call %s not found", clientMetadata.ModuleCallerDigest)
+	}
+
+	return callCtx.FnCall, nil
+}
+
+func (q *Query) CurrentServedDeps(ctx context.Context) (*ModDeps, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	callCtx, ok := q.clientCallContext[clientMetadata.ModuleCallerDigest]
+	if !ok {
+		return nil, fmt.Errorf("client call %s not found", clientMetadata.ModuleCallerDigest)
+	}
+	return callCtx.Deps, nil
+}
+
+func (q *Query) WithPipeline(name, desc string, labels []pipeline.Label) *Query {
+	q = q.Clone()
+	q.Pipeline = q.Pipeline.Add(pipeline.Pipeline{
+		Name:        name,
+		Description: desc,
+		Labels:      labels,
+	})
+	return q
+}
+
+func (q *Query) NewContainer(platform Platform) *Container {
+	return &Container{
+		Query:    q,
+		Platform: platform,
+	}
+}
+
+func (q *Query) NewSecret(name string) *Secret {
+	return &Secret{
+		Query: q,
+		Name:  name,
+	}
+}
+
+func (q *Query) NewHost() *Host {
+	return &Host{
+		Query: q,
+	}
+}
+
+func (q *Query) NewModule() *Module {
+	return &Module{
+		Query: q,
+	}
+}
+
+func (q *Query) NewContainerService(ctr *Container) *Service {
+	return &Service{
+		Query:     q,
+		Container: ctr,
+	}
+}
+
+func (q *Query) NewTunnelService(upstream dagql.Instance[*Service], ports []PortForward) *Service {
+	return &Service{
+		Query:          q,
+		TunnelUpstream: &upstream,
+		TunnelPorts:    ports,
+	}
+}
+
+func (q *Query) NewHostService(upstream string, ports []PortForward) *Service {
+	return &Service{
+		Query:        q,
+		HostUpstream: upstream,
+		HostPorts:    ports,
+	}
+}
+
+func (q *Query) IDDeps(ctx context.Context, baseDeps *ModDeps, id *idproto.ID) (*ModDeps, error) {
```

Comment:
I'm not totally grok-ing what's going on here and it's usage, comment would be helpful.

e.g. the usage in `ModuleObjectType.ConvertToSDKInput`, you set `baseDeps` to the module deps + itself, then you provide this `id` and append all of it's modules too (plus including the core mod via default deps). Are you duplicating modules in the list? Or are they non-overlapping sets of modules?

Reply:
Turns out we don't need the `baseDeps` arg anymore! :tada: It always felt fishy that it needed effort on both sides of the call.

In a nutshell, this is the secret sauce that allows IDs to bring-their-own-dependencies, so you can take an ID returned by a module and pass it to any other module where it'll be loaded using a throw-away `*dagql.Server` with all of the ID's dependencies installed.

This must have been fixed when I moved the dependency-tracking `Module *idproto.ID` field from `dagql.Class[T]` to `dagql.FieldSpec`, which allows fields like `impl` - which come from a different module than the class it's being installed onto - to correctly declare its dependent module.